src/main/java/gregtech/common/metatileentities/MetaTileEntityHeatExchanger.java

```diff
@@ -0,0 +1,62 @@
+package gregtech.common.metatileentities;
+
+import gregtech.api.capability.impl.PrimitiveRecipeLogic;
+import gregtech.api.metatileentity.MetaTileEntity;
+import gregtech.api.metatileentity.interfaces.IGregTechTileEntity;
+import gregtech.api.metatileentity.multiblock.IMultiblockPart;
+import gregtech.api.metatileentity.multiblock.RecipeMapMultiblockController;
+import gregtech.api.pattern.BlockPattern;
+import gregtech.api.pattern.FactoryBlockPattern;
+import gregtech.api.recipes.RecipeMaps;
+import gregtech.api.unification.material.Materials;
+import gregtech.client.renderer.ICubeRenderer;
+import gregtech.client.renderer.texture.Textures;
+import gregtech.common.blocks.BlockBoilerCasing.BoilerCasingType;
+import gregtech.common.blocks.BlockMetalCasing.MetalCasingType;
+import gregtech.common.blocks.MetaBlocks;
+
+import net.minecraft.util.ResourceLocation;
+
+import org.jetbrains.annotations.NotNull;
+
+public class MetaTileEntityHeatExchanger extends RecipeMapMultiblockController {
+
+    public MetaTileEntityHeatExchanger(ResourceLocation metaTileEntityId) {
+        super(metaTileEntityId, RecipeMaps.HEAT_EXCHANGER_RECIPES);
+        this.recipeMapWorkable = new PrimitiveRecipeLogic(this);
+    }
+
+    @Override
+    public MetaTileEntity createMetaTileEntity(IGregTechTileEntity iGregTechTileEntity) {
+        return new MetaTileEntityHeatExchanger(metaTileEntityId);
+    }
+
+    @Override
+    protected @NotNull BlockPattern createStructurePattern() {
+        return FactoryBlockPattern.start()
+                .aisle("CCC", "BCB", "ACA")
+                .aisle("CCC", "CDC", "ACA").setRepeatable(7)
+                .aisle("CCC", "BSB", "AEA")
+                .where('S', selfPredicate())
+                .where('A', frames(Materials.Steel))
+                .where('B', autoAbilities(false, false, false, false, false, true, false).setMinGlobalLimited(2)
```

Comment:
If you are not checking for muffler in `autoabilities`, it might be good to just override `hasMufflerMechanics`.